File: week-04/yc_1021_rotating_queue.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        ArrayDeque<Integer> dq = new ArrayDeque<>();
        for (int i = 1; i <= n; i++) {
            dq.addLast(i);
        }

        String[] targetStr = br.readLine().split(" ");
        int[] targets = Arrays.stream(targetStr).mapToInt(Integer::parseInt).toArray();

        int result = 0;

        for (int target : targets) {
            int index = 0;
            for (int val : dq) {
                if (val == target) break;
                index++;
            }

            int leftDistance = index;
            int rightDistance = dq.size() - index;

            if (leftDistance > rightDistance) {
                for (int i = 0; i < rightDistance; i++) {
                    dq.addFirst(dq.removeLast());
                    result++;
                }
            } else {
                for (int i = 0; i < leftDistance; i++) {
                    dq.addLast(dq.removeFirst());
                    result++;
                }
            }
            dq.removeFirst();
        }

        System.out.println(result);
    }
}

File: week-04/yc_1021_rotating_queue.cs
using System;
using System.Collections.Generic;

class Program {
    static void Main(string[] args)
    {
        var input = Console.ReadLine().Split();
        int n = int.Parse(input[0]);
        int m = int.Parse(input[1]);

        var dq = new LinkedList<int>();
        for (int i = 1; i <= n; i++)
        {
            dq.AddLast(i);
        }

        var targets = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);

        int result = 0;

        foreach (var target in targets)
        {
            int index = 0;
            var node = dq.First;
            while (node.Value != target)
            {
                node = node.Next;
                index++;
            }

            int leftDistance = index;
            int rightDistance = dq.Count - index;

            if (leftDistance > rightDistance)
            {
                for (int i = 0; i < rightDistance; i++)
                {
                    dq.AddFirst(dq.Last.Value);
                    dq.RemoveLast();
                    result++;
                }
            }
            else
            {
                for (int i = 0; i < leftDistance; i++)
                {
                    dq.AddLast(dq.First.Value);
                    dq.RemoveFirst();
                    result++;
                }
            }
            dq.RemoveFirst();
        }

        Console.WriteLine(result);
    }
}
